prover/src/logup_gkr/mod.rs

```diff
@@ -106,53 +111,72 @@ impl<E: FieldElement> EvaluatedCircuit<E> {
-                main_trace.read_main_frame(i, &mut main_frame);
-                periodic_values.fill_periodic_values_at(i, &mut periodic_values_row);
-                evaluator.build_query(&main_frame, &mut query);
-
-                evaluator.evaluate_query(
-                    &query,
-                    &periodic_values_row,
-                    log_up_randomness,
-                    &mut numerators,
-                    &mut denominators,
-                );
-                let input_gates_values: Vec<CircuitWire<E>> = numerators
-                    .iter()
-                    .zip(denominators.iter())
-                    .map(|(numerator, denominator)| CircuitWire::new(*numerator, *denominator))
-                    .collect();
-                input_gates_values
-            };
-
-            input_layer_wires.extend(wires_from_trace_row);
-        }
+            unsafe { uninit_vector(trace.main_segment().num_rows() * num_fractions) };
+        let num_cols = trace.main_segment().num_cols();
+        let num_oracles = evaluator.get_oracles().len();
+        let num_periodic_cols = periodic_values.num_columns();
+
+        batch_iter_mut!(
+            &mut input_layer_wires,
+            1024,
+            |batch: &mut [CircuitWire<E>], batch_offset: usize| {
+                let mut main_frame = EvaluationFrame::new(num_cols);
+                let mut query = vec![E::BaseField::ZERO; num_oracles];
+                let mut periodic_values_row = vec![E::BaseField::ZERO; num_periodic_cols];
+                let mut numerators = vec![E::ZERO; num_fractions];
+                let mut denominators = vec![E::ZERO; num_fractions];
+
+                let row_offset = batch_offset / num_fractions;
+                let batch_size = batch.len();
+                let num_rows_per_batch = batch_size / num_fractions;
+
+                for i in
+                    (0..trace.main_segment().num_rows()).skip(row_offset).take(num_rows_per_batch)
+                {
+                    let wires_from_trace_row = {
+                        trace.read_main_frame(i, &mut main_frame);
+                        periodic_values.fill_periodic_values_at(i, &mut periodic_values_row);
+                        evaluator.build_query(&main_frame, &mut query);
+
+                        evaluator.evaluate_query(
+                            &query,
+                            &periodic_values_row,
+                            log_up_randomness,
+                            &mut numerators,
+                            &mut denominators,
+                        );
+
+                        let input_gates_values: Vec<CircuitWire<E>> = numerators
+                            .iter()
+                            .zip(denominators.iter())
+                            .map(|(numerator, denominator)| {
+                                CircuitWire::new(*numerator, *denominator)
+                            })
+                            .collect();
```

Comment:
Allocating vectors here may add up as we do it for every row in the trace. As far as I can tell, we do this only so that we can copy the values to the `batch` immediately after we build the vector. Could we do it here? e.g., something like this:
```rust
let n = (i - row_offset) * num_fractions;
for (j, (numerator, denominator)) in
    numerators.iter().zip(denominators.iter()).enumerate()
{
    batch[n + j] = CircuitWire::new(*numerator, *denominator);
}
```
If this works, then we don't need to do `copy_from_slice()` at the of the iteration.

Reply:
Or maybe even like this:
```rust
let n = (i - row_offset) * num_fractions;
for ((wire, numerator), denominator) in
    batch[n..].iter_mut().zip(numerators.iter()).zip(denominators.iter())
{
    *wire = CircuitWire::new(*numerator, *denominator);
}
```

Reply:
Another option is to change `evaluate_query()` to something like:
```rust
fn evaluate_query<F, E>(
    &self,
    query: &[F],
    periodic_values: &[F],
    logup_randomness: &[E],
    wires: &mut [CircuitWire],
) where
    F: FieldElement<BaseField = Self::BaseField>,
    E: FieldElement<BaseField = Self::BaseField> + ExtensionOf<F>;
```
But maybe that's for another PR.

Reply:
Great idea! Implemented